src/System.Net.Security/src/System/Net/_SSPIWrapper.cs

```diff
@@ -0,0 +1,873 @@
+                    break;
+
+                case Interop.Secur32.ContextAttribute.LocalCertificate:
+                    handleType = typeof(SafeFreeCertContext);
+                    break;
+
+                case Interop.Secur32.ContextAttribute.IssuerListInfoEx:
+                    nativeBlockSize = Marshal.SizeOf<Interop.Secur32.IssuerListInfoEx>();
+                    handleType = typeof(SafeFreeContextBuffer);
+                    break;
+
+                case Interop.Secur32.ContextAttribute.ConnectionInfo:
+                    nativeBlockSize = Marshal.SizeOf<SslConnectionInfo>();
+                    break;
+
+                default:
+                    throw new ArgumentException(SR.Format(SR.net_invalid_enum, "ContextAttribute"), "contextAttribute");
+            }
+
+            SafeHandle SspiHandle = null;
+            object attribute = null;
+
+            try
+            {
+                byte[] nativeBuffer = new byte[nativeBlockSize];
+                errorCode = SecModule.QueryContextAttributes(securityContext, contextAttribute, nativeBuffer, handleType, out SspiHandle);
+                if (errorCode != 0)
+                {
+                    GlobalLog.Leave("Win32:QueryContextAttributes", "ERROR = " + ErrorDescription(errorCode));
+                    return null;
+                }
+
+                switch (contextAttribute)
+                {
+                    case Interop.Secur32.ContextAttribute.Sizes:
+                        attribute = new SecSizes(nativeBuffer);
+                        break;
+
+                    case Interop.Secur32.ContextAttribute.StreamSizes:
+                        attribute = new StreamSizes(nativeBuffer);
+                        break;
+
+                    case Interop.Secur32.ContextAttribute.Names:
+                        attribute = Marshal.PtrToStringUni(SspiHandle.DangerousGetHandle());
+                        break;
+
+                    case Interop.Secur32.ContextAttribute.PackageInfo:
+                        attribute = new SecurityPackageInfoClass(SspiHandle, 0);
+                        break;
+
+                    case Interop.Secur32.ContextAttribute.NegotiationInfo:
+                        unsafe
+                        {
+                            fixed (void* ptr = nativeBuffer)
+                            {
+                                attribute = new NegotiationInfoClass(SspiHandle, Marshal.ReadInt32(new IntPtr(ptr), NegotiationInfo.NegotiationStateOffest));
+                            }
+                        }
+                        break;
+
+                    case Interop.Secur32.ContextAttribute.ClientSpecifiedSpn:
+                        attribute = Marshal.PtrToStringUni(SspiHandle.DangerousGetHandle());
+                        break;
+
+                    case Interop.Secur32.ContextAttribute.LocalCertificate:
+                        goto case Interop.Secur32.ContextAttribute.RemoteCertificate;
```

Comment:
Why not just do this as a two-label case? The goto feels a little odd given that there's nothing else here and it's "goto next line".